Code changes to update references of the Identifiers

sarangan12 opened the pull request:
@paulvanbrenk @mhegazy Could you please review these code changes? I will update the PR with tests soon. 

<!--
Thank you for submitting a pull request!

Here's a checklist you might find useful.
[ ] There is an associated issue that is labelled
  'Bug' or 'Accepting PRs' or is in the Community milestone
[ ] Code is up-to-date with the `master` branch
[ ] You've successfully run `jake runtests` locally
[ ] You've signed the CLA
[ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
-->

Fixes #

msftclas commented:
Hi __@sarangan12__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

paulvanbrenk commented on src/compiler/types.ts:
```diff
@@ -2073,6 +2073,7 @@ namespace ts {
         /* @internal */ parent?: Symbol;        // Parent symbol
         /* @internal */ exportSymbol?: Symbol;  // Exported symbol associated with this symbol
         /* @internal */ constEnumOnlyModule?: boolean; // True if module contains only const enums or other modules with only const enums
+        hasReference?: boolean;
```
Add the `/* @internal */ marker

sarangan12 replied:
Done

sarangan12 commented:
Added code changes for handling local variables and parameters in functions, methods and constructors. Yet to add test cases. Updated the PR for early feedback.

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -14165,6 +14177,26 @@ namespace ts {
+
+        function checkUnusedIdentifiers(node: FunctionDeclaration | MethodDeclaration | ConstructorDeclaration): void {
+            if (!isSourceFileADefinitionFile(node)) {
+                for (const key in node.locals) {
```
`if (hasPropeorty(node.locals, key))`

sarangan12 replied:
Done

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -14165,6 +14177,26 @@ namespace ts {
+            if (!isSourceFileADefinitionFile(node)) {
+                for (const key in node.locals) {
+                    if (!node.locals[key].hasReference && node.locals[key].valueDeclaration && node.locals[key].valueDeclaration.kind) {
+                        if (compilerOptions.noUnusedLocals && node.locals[key].valueDeclaration.kind === SyntaxKind.VariableDeclaration) {
```
i would move the check to the top of the function, and bail out if it is false.

sarangan12 replied:
Done

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -8071,8 +8071,18 @@ namespace ts {
         }
 
+        function updateReferences(node: Identifier): void {
+            if (!isSourceFileADefinitionFile(node)) {
```
check the `noUnusedParameters` && `noUnusedLocals` before marking.

paulvanbrenk replied:
I would only check those when reporting.. the option check is probably more expensive than setting the property?

sarangan12 replied:
Marking it always and checking it only when reporting

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -8071,8 +8071,18 @@ namespace ts {
+            if (!isSourceFileADefinitionFile(node)) {
+                const symbol = getReferencedValueSymbol(node);
+                if (symbol) {
+                    symbol.hasReference = true;
```
i would store this on symbolLinks instead.

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -14165,6 +14177,26 @@ namespace ts {
+        }
+
+        function checkUnusedIdentifiers(node: FunctionDeclaration | MethodDeclaration | ConstructorDeclaration): void {
+            if (!isSourceFileADefinitionFile(node)) {
```
i would also check for node.body to avoid reporting errors for function overloads.

mhegazy replied:
i would change this condition to be `inAmbientContext`

mhegazy commented on src/compiler/commandLineParser.ts:
```diff
@@ -133,6 +133,16 @@ namespace ts {
+            description: Diagnostics.Variable_0_has_never_been_used,
+        },
+        {
+            name: "noUnusedParameters",
```
consider using just one flag

vladima commented on src/compiler/checker.ts:
```diff
@@ -14165,6 +14177,26 @@ namespace ts {
+            return getSourceFileOfNode(node).fileName.indexOf(".d.ts") > 0;
+        }
+
+        function checkUnusedIdentifiers(node: FunctionDeclaration | MethodDeclaration | ConstructorDeclaration): void {
```
it does not take block scoped locals into account

vladima commented on src/compiler/checker.ts:
```diff
@@ -14165,6 +14177,26 @@ namespace ts {
             }
         }
 
+        function isSourceFileADefinitionFile(node: Node): boolean {
```
can you use `isInAmbientContext` instead?

vladima commented on src/compiler/checker.ts:
```diff
@@ -8071,8 +8071,18 @@ namespace ts {
+
         function checkIdentifier(node: Identifier): Type {
             const symbol = getResolvedSymbol(node);
+            updateReferences(node);
```
now this will mark every value identifier even if they are not local/parameter. also given that you already have symbol on hands, why do you need to retrieve it again in `updateReferences`?

sarangan12 replied:
removed the update reference method

mhegazy commented:
a couple more notes:
- paramterProperties should always be marked as referenced, e.g.
```ts
class C {
    constructor(public a: number){}
}
```
- typeof queries should not trigger this marking, e.g.
```ts
function foo(a: number, b: typeof a) { b ++; } // a should be flagged as unused
```

billti commented on src/compiler/checker.ts:
```diff
@@ -14175,6 +14215,85 @@ namespace ts {
         }
 
+        function checkSourceFileADefinitionFile(node: Node): boolean {
+            return getSourceFileOfNode(node).fileName.indexOf(".d.ts") > 0;
```
Ideally this should check the name ends in ".d.ts", not just contains it somewhere (i.e. `filename.match(/\.d\.ts$/)`)

sarangan12 replied:
Changed the logic

mhegazy commented on src/compiler/commandLineParser.ts:
```diff
@@ -133,6 +133,16 @@ namespace ts {
+        {
+            name: "noUnusedParameters",
+            type: "boolean",
+            description: Diagnostics.Parameter_0_has_never_been_used
```
this should not use a the error message. it should use a message with the description. something like `Raise error for unused parameters`
